### PR TITLE
Update to add commutation rules to the LocalOptimizer engine

### PR DESCRIPTION
**Pull request by daisyrainsmith**
Update to add commutation rules to the optimizer engine.
1. Update _optimizer.py and _optimizer_test.py
- apply_commutation Boolean – tells LocalOptimizer whether to consider commutation rules
- Rearranged existing code to streamline and improve readability.
- Restructuring of optimizer to check if the next command is commutable with the current command, or if it may be part of a commutable circuit.
2. Update _basics.py and _basics_test.py
- Added attributes to do with commutation rules.
3. Update _command.py and _command_test.py
- Added commutation attributes and overlap function to streamline optimizer code.
4. Update _gates.py and _gates_test.py
- Added attributes to do with commutation rules.
- Updated Rxx, Ryy, Rzz to have interchangeable qubit indices.
5. Update _metagates.py
- Added commutable rule.
6. Addition of _relative_command.py and _relative_command_test.py
- Dummy command which can be used without an engine, used to define commutation rules.
7. Update restricted_gate_set.py
- By default set apply_commutation = True

### CODE REVIEW

**Comment by Takishima** [comment]
Thank you for your contribution, I will review it next week.
In the meantime, you could make sure that the docstring of all functions follow the current style used in ProjectQ (essentially the [Google docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) style)

File: projectq/cengines/_optimize.py
Comment (Takishima) [inline projectq/cengines/_optimize.py:224]: I would suggest renaming this method so that its functionality is more obvious. Perhaps something like `_can_merge_by_commutation` or something similar.

File: projectq/cengines/_optimize.py
Comment (Takishima) [inline projectq/cengines/_optimize.py]: Could you give an example for such a situation? I am not sure I understand that limitation.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:279]: Since a similar test could be designed and performed for `Rz`, `R` and `Ph`, consider making this test parametrised on these gates.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:302]: Same comment as above about making this test parameterised.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:325]: Same comment as above about making this test parameterised.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:347]: Same comment as above about making this test parameterised.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:369]: Same comment as above about making this test parameterised.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:391]: Same comment as above about making this test parameterised.

File: projectq/cengines/_optimize_test.py
Comment (Takishima) [inline projectq/cengines/_optimize_test.py:414]: Same comment as above about making this test parameterised.

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:98]: Are you sure about this circuit equivalence?

File: projectq/ops/_relative_command.py
Comment (Takishima) [inline projectq/ops/_relative_command.py]: I would simplify the docstrig for `_gate` by simply saying that it is the true gate class if `gate` is  a metagate.

File: projectq/ops/_relative_command.py
Comment (Takishima) [inline projectq/ops/_relative_command.py:81]:
```suggestion
        self._gate = self.gate._gate if isinstance(self.gate, ControlledGate) else gate
```

File: projectq/ops/_relative_command.py
Comment (Takishima) [reply projectq/ops/_relative_command.py:121]: Also, please use `isinstance()` (and `type(...)`) instead of relying on the `__class__` magic method.

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:243]: Missing `X`, `Ph` and `SqrtX`

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:258]: Missing `Y` and  `Ph`

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:273]: Missing `Z`, `T`, `S`, `Ph` and `R`

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:288]: Missing `X`, `Ph` and `SqrtX`

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:304]: Missing `Y` and `Ph`

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:321]: Missing `Z`, `T`, `S`, `Ph` and `R`

File: projectq/ops/_gates.py
Comment (Takishima) [inline projectq/ops/_gates.py:106]: Missing commutation with `Y`, `Ry` and `Ryy`

**Review by Takishima — CHANGES_REQUESTED** [review]
Many thanks for your contribution.

Here are some preliminary comments about the code. Mostly some minor adjustments and modifications to make the code better fit in the existing code base, as well as some remarks that might require some larger reworking of the existing PR.

Overall, I think this is great work and would be happy to merge that once we polish it a little more. Once you fix my current comments, I will go through the code once more.

File: projectq/cengines/_optimize.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize.py:216]: No I think it should just exit the inner loop. At this point we have determined that command_i cannot cancel, merge or commute with command_i+x+1, so it should stop trying to looking at command_i’s relationship with future commands and move on to index i+1, i.e. move to the next iteration of the outer while loop.

File: projectq/cengines/_optimize.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize.py:140]: Done.

File: projectq/cengines/_optimize.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize.py]: In fact I got rid of _i_x_com completely because I realised it wasn’t being used to change any logic.

File: projectq/ops/_command.py
Comment (daisyrainsmith) [reply projectq/ops/_command.py:175]: I have changed is_commutable to work with enums. . is_commutable now returns a value such as: Commutable.COMMUTABLE.value. I chose to keep it returning an integer (the value of COMMUTABLE) rather than an enum, because then in my test files I can keep the nice simple readability of ‘assert gate1.is_commutable(gate2)’ and ‘assert not gate2.is_commutable(gate3)’. For the third option (a commutable circuit) you use: ‘assert gate3.is_commutable(gate4) == Commutable.MAYBE_COMMUTABLE

So this method already returns an int

File: projectq/ops/_command.py
Comment (daisyrainsmith) [reply projectq/ops/_command.py:156]: Done. is_commutable now returns a value such as: Commutable.COMMUTABLE.value. I chose to keep it returning an integer (the value of COMMUTABLE) rather than an enum, because then in my test files I can keep the nice simple readability of ‘assert gate1.is_commutable(gate2)’ and ‘assert not gate2.is_commutable(gate3)’. For the third option (commutable circuit) you use: ‘assert gate3.is_commutable(gate4) == Commutable.MAYBE_COMMUTABLE

File: projectq/ops/_gates.py
Comment (daisyrainsmith) [reply projectq/ops/_gates.py:78]: Done. However I had to define the XGate commutable gates exclusively within a get_method. This is a bit of a messy fix, but the alternative is to look into defining commutation rules once all gates are defined. Which is a project I don’t want to look into right now.

File: projectq/ops/_basics.py
Comment (daisyrainsmith) [reply projectq/ops/_basics.py:105]: I don’t think this would work because we want to be able to choose different values for commutable_circuit_list and commutable_gates for gates wrapped in a ControlledGate. 
For instance I have defined a get_commutable_circuit_list function in the XGate class which takes the parameter n as the number of control qubits and can return different commutable circuit lists depending on n. As far as I can tell the only way the optimizer can access the number of controls is at run time, by evaluating len(self.control_qubits) (see _command.py::Command::is_commutable). Therefore the _commutable_gates and _commutable_circuit_list attributes need to be instance attributes.

File: projectq/ops/_basics.py
Comment (daisyrainsmith) [reply projectq/ops/_basics.py]: see previous reply

Comment (daisyrainsmith) [reply projectq/ops/_basics.py]: See previous reply.

File: projectq/cengines/_optimize.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize.py:278]: No I think it should just exit the inner loop. At this point we have established that command_i cannot cancel, merge or commute with command_i+x+1, so it should stop trying to looking at command_i’s relationship with future commands and move on to index i+1, i.e. move to the next iteration of the outer while loop.

File: projectq/cengines/_optimize.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize.py:216]: No I think it should just exit the inner loop. At this point we have determined that command_i cannot cancel, merge or commute with command_i+x+1, so it should stop trying to looking at command_i’s relationship with future commands and move on to index i+1, i.e. move to the next iteration of the outer while loop.

File: projectq/cengines/_optimize.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize.py]: I think I just meant that this function does not take into account commutable_circuits. I can't think of a situation where this would be relevant actually so I'll just get rid of this comment.

File: projectq/ops/_gates.py
Comment (daisyrainsmith) [reply projectq/ops/_gates.py:102]: I got this commutation rule from this paper:
https://arxiv.org/abs/1710.07345#:~:text=Automated%20optimization%20of%20large%20quantum%20circuits%20with%20continuous%20parameters,-Yunseong%20Nam%2C%20Neil&text=In%20particular%2C%20we%20provide%20better,of%20the%20underlying%20quantum%20algorithms.
(Figure 5)

File: projectq/cengines/_optimize.py
Comment (Takishima) [reply projectq/cengines/_optimize.py:278]: Oh ok that was my misunderstanding.

File: projectq/ops/_basics.py
Comment (Takishima) [reply projectq/ops/_basics.py:105]: Ok, makes sense.

File: projectq/ops/_gates.py
Comment (Takishima) [reply projectq/ops/_gates.py:78]: If not already done, could you add this remark as a comment somewhere in the code (e.g. maybe at the top of the file), so that we know what a possible next step could be?

File: projectq/ops/_command.py
Comment (Takishima) [reply projectq/ops/_command.py:156]: Instead of an enum class deriving from `Enum`, why don't you consider deriving from `IntEnum`?
The latter gives you conversion to `int` without needing to access the `value` attribute.

File: projectq/ops/_command.py
Comment (Takishima) [reply projectq/ops/_command.py:175]: See comment https://github.com/ProjectQ-Framework/ProjectQ/pull/386#discussion_r607637535

File: projectq/cengines/_optimize.py
Comment (Takishima) [reply projectq/cengines/_optimize.py:216]: Ok.

File: projectq/ops/_command.py
Comment (daisyrainsmith) [reply projectq/ops/_command.py:156]: Done. As will be seen in the next commit.

File: projectq/ops/_gates.py
Comment (daisyrainsmith) [reply projectq/ops/_gates.py:78]: Done. As will be seen in next commit.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:323]: Done. As will be seen in next commit.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:346]: Done. As will be seen in next commit.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:369]: Done. As will be seen in next commit.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:391]: Done. As will be seen in next commit.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:413]: Done. As will be seen in next commit.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:435]: Done. As will be seen in next commit. Ph was missed out because it commutes through the control (of the CNOT) unlike the Rz and R. (Although in theory the Rz and R should commute through the control as well but this would require a new update.

File: projectq/cengines/_optimize_test.py
Comment (daisyrainsmith) [reply projectq/cengines/_optimize_test.py:458]: Done. As will be seen in next commit.

File: projectq/ops/_gates.py
Comment (daisyrainsmith) [reply projectq/ops/_gates.py:110]: Done. As will be seen in next commit. I haven't listed Y as a gate which commutes with YGate (or put in logic for any gates to commute with themselves) because I can't think of any situations where that would be useful.

File: projectq/ops/_command.py
Comment (daisyrainsmith) [reply projectq/ops/_command.py:175]: Done. As will be seen in next commit.

**Comment by daisyrainsmith** [comment]
Hi Takishima, I was just wondering if you've had a chance to look at the latest commits? I hope you haven't been waiting on me for something because I was unaware. Thanks for all your help so far!
Daisy.

**Comment by Takishima** [comment]
Sorry, I was a bit busy with the other PRs and some other projects at work. I will look at your latest changes in the coming days.

Be aware, I will not merge any PR until #395 gets merged since that one introduces a few important changes in the way the project is configured.

**Comment by daisyrainsmith** [comment]
No worries, just wanted to check we weren't unwittingly both waiting on each other. Thanks, Daisy.

**Comment by Takishima** [comment]
Dear @daisyrainsmith,

I have taken a deeper look at your contribution today and things look good for the most part. There are still a few places where I would like to see some cleanup like in some of your test functions that are a bit overcrowded for a lack of a better word.

But more importantly, the current test suite does not seem to cover all the code you have improved upon. I do not have time right now to look into those places in more details to see if they are important or not, but here is the result of me running `pytest-cov` on the latest commit to date.
Note in particular the cases for `projectq/ops/_relative_command.py` and `projectq/cengines/_optimize.py` where I think most of your modifications lie.

```
---------- coverage: platform darwin, python 3.9.5-final-0 -----------
Name                                                          Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------------------
projectq/__init__.py                                              3      0   100%
projectq/_version.py                                              2      0   100%
projectq/backends/__init__.py                                     8      0   100%
projectq/backends/_aqt/__init__.py                                1      0   100%
projectq/backends/_aqt/_aqt.py                                  157      0   100%
projectq/backends/_aqt/_aqt_http_client.py                      109      0   100%
projectq/backends/_awsbraket/__init__.py                          2      0   100%
projectq/backends/_awsbraket/_awsbraket.py                      158      0   100%
projectq/backends/_awsbraket/_awsbraket_boto3_client.py         165      0   100%
projectq/backends/_circuits/__init__.py                           4      0   100%
projectq/backends/_circuits/_drawer.py                           89      0   100%
projectq/backends/_circuits/_drawer_matplotlib.py                82      0   100%
projectq/backends/_circuits/_plot.py                            162      0   100%
projectq/backends/_circuits/_to_latex.py                        364      5    99%   60-61, 79-81
projectq/backends/_ibm/__init__.py                                1      0   100%
projectq/backends/_ibm/_ibm.py                                  169      0   100%
projectq/backends/_ibm/_ibm_http_client.py                      171      0   100%
projectq/backends/_printer.py                                    47      1    98%   106
projectq/backends/_resource.py                                   88      0   100%
projectq/backends/_sim/__init__.py                                2      0   100%
projectq/backends/_sim/_classical_simulator.py                  105      0   100%
projectq/backends/_sim/_pysim.py                                243      1    99%   27
projectq/backends/_sim/_simulator.py                            135      3    98%   39-41
projectq/cengines/__init__.py                                    13      0   100%
projectq/cengines/_basicmapper.py                                38      0   100%
projectq/cengines/_basics.py                                     61      0   100%
projectq/cengines/_cmdmodifier.py                                 9      0   100%
projectq/cengines/_ibm5qubitmapper.py                            77      0   100%
projectq/cengines/_linearmapper.py                              318      0   100%
projectq/cengines/_main.py                                       96      2    98%   196-197
projectq/cengines/_manualmapper.py                               16      0   100%
projectq/cengines/_optimize.py                                  266     18    93%   75-76, 135-137, 166-168, 209, 274-275, 375-377, 380-382, 386, 494, 549
projectq/cengines/_replacer/__init__.py                           3      0   100%
projectq/cengines/_replacer/_decomposition_rule.py               12      0   100%
projectq/cengines/_replacer/_decomposition_rule_set.py           31      1    97%   70
projectq/cengines/_replacer/_replacer.py                         68      0   100%
projectq/cengines/_swapandcnotflipper.py                         59      0   100%
projectq/cengines/_tagremover.py                                 13      0   100%
projectq/cengines/_testengine.py                                 64      0   100%
projectq/cengines/_twodmapper.py                                335      0   100%
projectq/libs/__init__.py                                         0      0   100%
projectq/libs/hist/__init__.py                                    2      0   100%
projectq/libs/hist/_histogram.py                                 32      0   100%
projectq/libs/math/__init__.py                                    2      0   100%
projectq/libs/math/_constantmath.py                              59      0   100%
projectq/libs/math/_default_rules.py                             90      0   100%
projectq/libs/math/_gates.py                                    174      0   100%
projectq/libs/math/_quantummath.py                              188      0   100%
projectq/libs/revkit/__init__.py                                  3      0   100%
projectq/libs/revkit/_control_function.py                        28     22    21%   55-69, 81-112, 120-121
projectq/libs/revkit/_permutation.py                             20     14    30%   48-51, 61-84, 92-94
projectq/libs/revkit/_phase.py                                   28     22    21%   59-73, 83-114, 122-123
projectq/libs/revkit/_utils.py                                    3      2    33%   25-26
projectq/meta/__init__.py                                         8      0   100%
projectq/meta/_compute.py                                       156      0   100%
projectq/meta/_control.py                                        38      0   100%
projectq/meta/_dagger.py                                         37      0   100%
projectq/meta/_dirtyqubit.py                                      6      0   100%
projectq/meta/_logicalqubit.py                                    8      0   100%
projectq/meta/_loop.py                                           89      0   100%
projectq/meta/_util.py                                           10      0   100%
projectq/ops/__init__.py                                         12      0   100%
projectq/ops/_basics.py                                         172      1    99%   146
projectq/ops/_command.py                                        114      0   100%
projectq/ops/_gates.py                                          212      1    99%   386
projectq/ops/_metagates.py                                       87      0   100%
projectq/ops/_qaagate.py                                          8      0   100%
projectq/ops/_qftgate.py                                          5      0   100%
projectq/ops/_qpegate.py                                          7      0   100%
projectq/ops/_qubit_operator.py                                 235      0   100%
projectq/ops/_relative_command.py                                31     15    52%   77, 86-100
projectq/ops/_shortcuts.py                                        8      0   100%
projectq/ops/_state_prep.py                                      15      0   100%
projectq/ops/_time_evolution.py                                  68      0   100%
projectq/ops/_uniformly_controlled_rotation.py                   56      0   100%
projectq/setups/__init__.py                                       0      0   100%
projectq/setups/aqt.py                                           27      0   100%
projectq/setups/awsbraket.py                                     22      0   100%
projectq/setups/decompositions/__init__.py                        2      0   100%
projectq/setups/decompositions/amplitudeamplification.py         25      0   100%
projectq/setups/decompositions/arb1qubit2rzandry.py              86      0   100%
projectq/setups/decompositions/barrier.py                         8      0   100%
projectq/setups/decompositions/carb1qubit2cnotrzandry.py        106      0   100%
projectq/setups/decompositions/cnot2cz.py                        14      0   100%
projectq/setups/decompositions/cnot2rxx.py                       24      0   100%
projectq/setups/decompositions/cnu2toffoliandcu.py               31      0   100%
projectq/setups/decompositions/crz2cxandrz.py                    16      0   100%
projectq/setups/decompositions/entangle.py                       12      0   100%
projectq/setups/decompositions/globalphase.py                     9      0   100%
projectq/setups/decompositions/h2rx.py                           18      0   100%
projectq/setups/decompositions/ph2r.py                           13      0   100%
projectq/setups/decompositions/phaseestimation.py                23      0   100%
projectq/setups/decompositions/qft2crandhadamard.py              15      0   100%
projectq/setups/decompositions/qubitop2onequbit.py               24      0   100%
projectq/setups/decompositions/r2rzandph.py                      12      0   100%
projectq/setups/decompositions/rx2rz.py                          16      0   100%
projectq/setups/decompositions/ry2rz.py                          17      0   100%
projectq/setups/decompositions/rz2rx.py                          26      0   100%
projectq/setups/decompositions/sqrtswap2cnot.py                  17      0   100%
projectq/setups/decompositions/stateprep2cnot.py                 51      0   100%
projectq/setups/decompositions/swap2cnot.py                      13      0   100%
projectq/setups/decompositions/time_evolution.py                 63      0   100%
projectq/setups/decompositions/toffoli2cnotandtgate.py           28      0   100%
projectq/setups/decompositions/uniformlycontrolledr2cnot.py      65      0   100%
projectq/setups/default.py                                        7      0   100%
projectq/setups/grid.py                                          59      0   100%
projectq/setups/ibm.py                                           42      0   100%
projectq/setups/linear.py                                        59      0   100%
projectq/setups/restrictedgateset.py                             95      0   100%
projectq/setups/trapped_ion_decomposer.py                        45      0   100%
projectq/tests/__init__.py                                        0      0   100%
projectq/types/__init__.py                                        1      0   100%
projectq/types/_qubit.py                                         72      0   100%
-------------------------------------------------------------------------------------------
TOTAL                                                          6920    108    98%
```

**Comment by daisyrainsmith** [comment]
Hi Takishima, 

I have submitted a new Pull Request because it was too difficult to rebase to I have submitted a simple merge between my commutation update and your latest develop branch. 

If that's alright, please close this pull request and we'll move to the other one. 

Thanks, 

Daisy.

**Comment by Takishima** [comment]
Ok, I'll close this one then.

**Pull request closed without merging**
